Feature/simple computer

serpient opened the pull request:
# What
- Computer player chooses a random empty spot on board
- Game settings can choose a local or computer as opponent
- Updated validation for the 2 different game settings (board size and opponent type)
- Based on opponent type, switches the play functions for each player type (human or computer playing)
- Tweaked printing to show computer pick message before prompting for users.
- Rebased from refactor branch which pulled in the new custom board size branch commits.
- Move around functions so that the public function is at the top, and implementation functions are ordered below
- Extracted human vs computer turn functions and corresponding input validation into its own modules. Hopefully, this is a good use of Single Responsibility Principle. 

# Questions
- Most of the functions in TicTacToe.ex is not tested because they call other functions in a loop. How do I test for that condition???

# Screenshots
![gif](https://media.giphy.com/media/MC8smtRsz4ajjbqTG1/giphy.gif)

serpient committed:
picks chooses random empty space

serpient committed:
add computer_turn- picks space & updates board

serpient committed:
add game setting validation fns

serpient committed:
integrate validation fns into game_start; extract human vs computer turn play

serpient committed:
use updated atoms for computer vs local

serpient committed:
update validation fns for board_size & opponent_type

serpient committed:
create game_settings to set opponent and board size

serpient committed:
implement game settings call in TTT

serpient committed:
update opponent type message; add error as param

serpient committed:
add symbol for :ai

serpient committed:
use correct player symbols in game messages

serpient committed:
mix format

serpient committed:
change "spec" to "state"

serpient committed:
rename fn to get_ for game_settings

serpient committed:
update turn fn name to get_*_move

serpient committed:
adjusting printing of computer message

- fixed bugs where it wasnt printing and clearing the console correctly during local vs local and error printing

serpient committed:
clean up message printing logic

serpient committed:
extract computer message into own fn

serpient committed:
add computer player gif to readme

serpient committed:
add :ai to be checked for in win fn

serpient committed:
add :ai to be checked for in win fn

serpient committed:
extract clear console logic

serpient committed:
rename fn to get_computer_message for clarity

serpient committed:
pull out print_board from get_human_move

- unrelated to getting a humans move
- moved in front of the fn call instead

serpient committed:
extract human and computer move logic

serpient committed:
move related input processing into relevant module

serpient committed:
move around fns to expose fn vs implementation fns

serpient committed:
bare values; clean up case/cond

jayden commented on lib/check_for_wins.ex:
```diff
@@ -22,12 +20,17 @@ defmodule CheckForWins do
+  def has_winning_row?(all_rows) do
+    all_rows
+    |> Enum.any?(fn row ->
+      is_row_a_win?(row, :player)
```
It seems odd that you are checking for all three player types here. Instead of passing in a symbol to check against, we could determine if it's a row win if every symbol in the row is the same.

serpient committed:
using for-do instead of map for clarity

jayden commented on lib/game_IO.ex:
```diff
@@ -9,35 +9,55 @@ defmodule GameIO do
 
+  def get_computer_message(key, value) do
+    computer_player_message = %{
+      chooses_spot: ~s(\n\nComputer chooses spot #{value}.\n\n\n)
```
Are you expecting to support other messages here soon? Curious why you chose to use a map here instead of returning an interpolated string.

serpient replied:
I had spoken with another apprentice who had a large set of computer messages, so I was preparing to do the same. But if there's no further features, ill just have it return the single string 👍

jayden commented on lib/game_settings.ex:
```diff
@@ -0,0 +1,59 @@
+  def handle_board_size(result) do
+    case result do
+      {:ok, message} -> message
+      {:error, message} -> GameSettings.get_board_size(message)
```
I notice that you sometimes prepend the module name (within the same module) and sometimes you don't. For consistency's sake, we should pick one and stick to it 😺 

The same comment applies for cases when a parameterless function is invoked (`do_something` vs `do_something()`).

jayden commented on lib/printable_board.ex:
```diff
@@ -28,19 +44,18 @@ defmodule PrintableBoard do
-        board_data
-      )
-    end)
+    cond do
```
Is there an advantage to using `cond` for a simple conditional like this vs using an `if-else`?

serpient replied:
no strong reason -- I usually prefer `cond-do` cause it seems more easily readable. I'll convert to a `if-else`

jayden commented on README.md:
```diff
@@ -26,6 +26,8 @@ If no win -- continue
 - Ability to play against a computer player. The computer should just choose random moves. 
 - Ability to choose between a 3x3 and 4x4 board.
 
+![gif](https://media.giphy.com/media/MC8smtRsz4ajjbqTG1/giphy.gif)
```
You may want to consider updating the README and gif since we no longer run the game using `TTT.play()`

jayden commented on test/computer_player_test.exs:
```diff
@@ -0,0 +1,39 @@
+defmodule Computer_Player_Test do
```
Is it conventional to use PascalCase and underscores for test module names?

jayden commented on lib/game_IO.ex:
```diff
@@ -9,35 +9,55 @@ defmodule GameIO do
       wins_game: ~s(Wins! \n\n),
       nil: ~s(Select a numbered spot: ),
```
Using `nil` as a key name seems a bit odd and can be confusing since it is also a value. Maybe renaming this to something more meaningful like `select_input` would be more clear.

serpient replied:
I've changed it to have a default key, which is really the behavior I wanted

serpient committed:
update is_row_a_win to check for matching non-empty symbol

serpient committed:
consistency in fn call and module name use

serpient committed:
convert conditional to if-else

serpient committed:
update readme with new game start; remove old gif

serpient committed:
rename test modules to be PascalCase; small cleanup

serpient committed:
update game_io messages with default value

serpient committed:
update comp msg to return single string instead of map+key

jayden reviewed (commented):
Thanks for addressing my comments. Admittedly, I'm not very familiar with Elixir so I'll defer to Amanda to approve this PR.

Something I wanted to bring up for you to consider is how granular a published commit should be. During development, it can be beneficial and good practice to commit early and often; but once published, super granular commits can make the git history difficult to read and not incredibly useful in a team setting.

Here's a pretty good article on granularity: http://gitforteams.com/resources/commit-granularity.html

aaizuss commented on lib/board_state.ex:
```diff
@@ -34,31 +39,37 @@ defmodule BoardState do
+    |> Enum.map(fn {_key, value} -> GameIO.get_player_token(value) end)
+  end
+
+  def get_empty_values(state) do
```
What's your reasoning for defining this function in this module as opposed to in the `Board` module?

serpient replied:
Only because it felt like the values I was iterating through felt very closely tied to board state. But that doesn't feel like a very strong reason. Shall I just move into Board?

aaizuss replied:
It looks like this function only has to do with the board, not any other part of state. Maybe I'm missing something, but it seems like the board module should know about its empty values.

serpient commented:
@jayden  thanks for reviewing! I'll read over the commit article -- I had thought that committing often was a good thing but didn't realize there were certain cases when it can be good/base.

aaizuss commented on lib/board_state.ex:
```diff
@@ -34,31 +39,37 @@ defmodule BoardState do
 
-  def update_board(position_to_update, board_state, player_symbol) do
-    board(board_state)
+  def update_board(position_to_update, state, player_symbol) do
```
I'm a bit confused about the relationship between the `BoardState` and `Board` modules. Last Friday, I thought it made sense, but now I'm concerned by how much BoardState knows about the way Board is implemented. For example, I would imagine that `Board` would have a function called `mark_space` or something like that which would be implemented using `Map.replace`, and then maybe BoardState takes care of updating state based on that? I thought BoardState was a higher level abstraction that uses Board, but it looks like Board actually depends on BoardState. It might be useful to draw a dependency diagram for all of your modules to see how effective your abstractions are.

serpient replied:
Currently, `Board` uses `BoardState`. 

for example, `Board` has a `handle_board_update` which calls `BoardState` to do the actual updating of the board and returns a new state. I thought this abstraction was helpful so that if any State implementation changes, its handled in one spot?

aaizuss replied:
It seems like BoardState should depend on Board, not the other way around. Maybe the `Board` module should be doing the updating, and `BoardState` should handle the update.  I like that you're thinking about what happens if the implementation of state changes, but the way this function works, it's actually coupled to the implementation of `Board`. You use `Map.replace!` which means that if a board is no longer implemented as a map, this function won't work. Now that I write this out, I think the issue is that there are multiple places that know how a board is implemented - `BoardState` and `Board`.  The `new_board` function should also be moved to the `Board` module. What if this module was called `GameState` (is opponent_type referring to the next player or the current player?) Sorry if this comment makes you more confused - hopefully discussing it will help clarify your intentions and our understanding

serpient replied:
yeah.... I am confused -- I think looking over this in person would help so we can discuss it in real time. I'm meeting / pairing with @jayden tomorrow so maybe this is one of the topics that we can discuss

serpient committed:
take apart gameSettings into 2 modules

- game settings really had 2 different logic changes. Handling board size settings and opponent type settings

serpient committed:
update game setting reference in TTT

serpient committed:
re-order game settings fn per usage
